File: ReSharper.FSharp/src/Psi.FSharp/src/FSharpProjectOptionsBuilder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using JetBrains.Annotations;
using JetBrains.Application.Components;
using JetBrains.Platform.MsBuildModel;
using JetBrains.Platform.ProjectModel.FSharp.Properties;
using JetBrains.ProjectModel;
using JetBrains.ProjectModel.Model2.Assemblies.Interfaces;
using JetBrains.ProjectModel.ProjectsHost;
using JetBrains.ProjectModel.ProjectsHost.MsBuild;
using JetBrains.ProjectModel.ProjectsHost.SolutionHost;
using JetBrains.ProjectModel.Properties.Managed;
using JetBrains.ReSharper.PsiGen.Util;
using JetBrains.ReSharper.Resources.Shell;
using JetBrains.Util;
using Microsoft.FSharp.Collections;
using Microsoft.FSharp.Compiler;
using Microsoft.FSharp.Compiler.SourceCodeServices;

namespace JetBrains.ReSharper.Psi.FSharp
{
  [SolutionComponent]
  public class FSharpProjectOptionsBuilder
  {
    private readonly ISolution mySolution;
    private readonly MsBuildProjectHost myMsBuildHost;

    public FSharpProjectOptionsBuilder(ISolution solution)
    {
      mySolution = solution;
      myMsBuildHost = solution.ProjectsHostContainer().GetComponent<MsBuildProjectHost>();
    }

    /// <summary>
    /// Build project options for FSharp.Compiler.Service. Options for referenced projects are not created.
    /// Use AddReferencedProjects to add referenced projects options.
    /// </summary>
    public FSharpProjectOptions BuildWithoutReferencedProjects(Guid projectGuid)
    {
      var project = mySolution.GetProjectByGuid(projectGuid);
      Assertion.AssertNotNull(project, "project != null");
      var buildSettings = project.ProjectProperties.BuildSettings as IManagedProjectBuildSettings;
      var configuration =
        project.ProjectProperties.ActiveConfigurations.Configurations.SingleItem() as IManagedProjectConfiguration;
      var projectOptions = new List<string>
      {
        "--out:" + project.GetOutputFilePath(project.GetCurrentTargetFrameworkId()),
        "--simpleresolution",
        "--noframework",
        "--debug:full",
        "--debug+",
        "--optimize-",
        "--tailcalls-",
        "--fullpaths",
        "--flaterrors",
        "--highentropyva+",
        "--target:" + GetOutputTarget(buildSettings),
        "--platform:anycpu", // todo
      };

      if (configuration != null)
      {
        projectOptions.addAll(SplitDefines(configuration.DefineConstants).Convert(s => "--define:" + s));

        var doc = configuration.DocumentationFile;
        if (!doc.IsNullOrWhitespace()) projectOptions.Add("--doc:" + doc);

        var nowarn = FixNoWarn(configuration.NoWarn);
        if (!nowarn.IsNullOrWhitespace()) projectOptions.Add("--nowarn:" + nowarn);

        projectOptions.Add("--warn:" + configuration.WarningLevel);

        // todo: add F# specific options like 'warnon'
      }

      projectOptions.addAll(GetReferencedPathsOptions(project));

      var projectFileNames = new List<string>();
      var projectMark = project.GetProjectMark().NotNull();
      myMsBuildHost.mySessionHolder.Execute(session =>
      {
        session.EditProject(projectMark, editor =>
        {
          // obtains all items in a right order directly from msbuild
          foreach (var item in editor.Items)
          {
            if (!BuildAction.GetOrCreate(item.ItemType()).IsCompile()) continue;

            var path = FileSystemPath.TryParse(item.EvaluatedInclude);
            if (path.IsEmpty) continue;

            var actualPath = EnsureAbsolute(path, projectMark.Location.Directory);
            projectFileNames.Add(actualPath.FullPath);
          }
        });
      });

      return new FSharpProjectOptions(
        project.ProjectFileLocation.FullPath,
        projectFileNames.ToArray(),
        projectOptions.ToArray(),
        referencedProjects: EmptyArray<Tuple<string, FSharpProjectOptions>>.Instance,
        isIncompleteTypeCheckEnvironment: false,
        useScriptResolutionRules: false,
        loadTime: DateTime.Now,
        unresolvedReferences: null,
        originalLoadReferences: FSharpList<Tuple<Range.range, string>>.Empty, // todo: check this
        extraProjectInfo: null
      );
    }

    private static string GetOutputTarget([CanBeNull] IManagedProjectBuildSettings buildSettings)
    {
      if (buildSettings != null && buildSettings.OutputType == ProjectOutputType.CONSOLE_EXE)
        return "exe";
      return "library"; // todo: any other targets for F#?
    }

    [NotNull]
    private static FileSystemPath EnsureAbsolute([NotNull] FileSystemPath path,
      [NotNull] FileSystemPath projectDirectory)
    {
      var relativePath = path.AsRelative();
      return relativePath != null
        ? projectDirectory.Combine(relativePath)
        : path;
    }

    [NotNull]
    private static string FixNoWarn([CanBeNull] string noWarn)
    {
      return SplitAndTrim(noWarn).Join(",");
    }

    [NotNull]
    private static string[] SplitDefines([CanBeNull] string definesString)
    {
      return string.IsNullOrEmpty(definesString)
        ? EmptyArray<string>.Instance
        : SplitAndTrim(definesString).ToArray();
    }

    [NotNull]
    private static IEnumerable<string> SplitAndTrim([CanBeNull] string strings)
    {
      return strings?.Split(';', ',', ' ').Select(x => x.Trim()).Where(s => !s.IsEmpty()) ??
             EmptyList<string>.Instance;
    }

    /// <summary>
    /// Current configuration defines, e.g. TRACE or DEBUG. Used in FSharpLexer.
    /// </summary>
    [NotNull]
    public string[] GetDefines(Guid projectGuid)
    {
      using (ReadLockCookie.Create())
      {
        var project = mySolution.GetProjectByGuid(projectGuid);
        Assertion.AssertNotNull(project, "project != null");
        var configuration =
          project.ProjectProperties.ActiveConfigurations.Configurations.SingleItem() as IManagedProjectConfiguration;
        return SplitDefines(configuration?.DefineConstants);
      }
    }

    [NotNull]
    private IEnumerable<string> GetReferencedPathsOptions(IProject project)
    {
      // todo: provide path to dll for unloaded projects?
      var framework = project.GetCurrentTargetFrameworkId();
      var refProjectsOutputs = project.GetReferencedProjects(framework)
        .Select(p => p.GetOutputFilePath(p.GetCurrentTargetFrameworkId()));
      var refAssembliesPaths = project.GetAssemblyReferences(framework)
        .Select(a => a.ResolveResultAssemblyFile().Location);
      return refProjectsOutputs.Concat(refAssembliesPaths).Select(a => "-r:" + a.FullPath);
    }

    [NotNull]
    public Dictionary<Guid, FSharpProjectOptions> AddReferencedProjects(
      [NotNull] IDictionary<Guid, FSharpProjectOptions> optionsDict)
    {
      var newOptions = new Dictionary<Guid, FSharpProjectOptions>();
      foreach (var projectGuidAndOptions in optionsDict)
      {
        var project = mySolution.GetProjectByGuid(projectGuidAndOptions.Key);
        Assertion.AssertNotNull(project, "project != null");
        var projectOptions = projectGuidAndOptions.Value;
        if (!newOptions.ContainsKey(project.Guid))
          newOptions[project.Guid] = AddReferencedProjects(project, projectOptions, optionsDict, newOptions);
      }

      return newOptions;
    }

    [NotNull]
    private FSharpProjectOptions AddReferencedProjects([NotNull] IProject project, FSharpProjectOptions projectOptions,
      [NotNull] IDictionary<Guid, FSharpProjectOptions> optionsDict,
      [NotNull] IDictionary<Guid, FSharpProjectOptions> newOptions)
    {
      // do not transitively add referenced projects (same as in other FSharp tools)
      var referencedProjects = project.GetReferencedProjects(project.GetCurrentTargetFrameworkId(), transitive: false);
      var referencedProjectsOptions = new List<Tuple<string, FSharpProjectOptions>>();
      foreach (var referencedProject in referencedProjects)
      {
        if (!(referencedProject.ProjectProperties is FSharpProjectProperties) ||
            newOptions.ContainsKey(project.Guid)) continue;

        var referencedProjectGuid = referencedProject.Guid;
        var referencedProjectOptions = optionsDict.ContainsKey(referencedProjectGuid)
          ? optionsDict[referencedProjectGuid]
          : BuildWithoutReferencedProjects(referencedProjectGuid);

        var fixedOptions = AddReferencedProjects(referencedProject, referencedProjectOptions, optionsDict, newOptions);
        newOptions[referencedProjectGuid] = fixedOptions;
        var outPath = referencedProject.GetOutputFilePath(referencedProject.GetCurrentTargetFrameworkId()).FullPath;
        referencedProjectsOptions.Add(Tuple.Create(outPath, fixedOptions));
      }

      return new FSharpProjectOptions(
        projectOptions.ProjectFileName,
        projectOptions.ProjectFileNames,
        projectOptions.OtherOptions,
        referencedProjectsOptions.ToArray(),
        projectOptions.IsIncompleteTypeCheckEnvironment,
        projectOptions.UseScriptResolutionRules,
        projectOptions.LoadTime,
        projectOptions.UnresolvedReferences,
        projectOptions.OriginalLoadReferences,
        projectOptions.ExtraProjectInfo);
    }
  }
}

File: ReSharper.FSharp/src/FSharp.Common/ProjectOptions/FSharpProjectOptionsBuilder.fs
namespace JetBrains.ReSharper.Plugins.FSharp.Common.ProjectOptions

open System
open System.Collections.Generic
open JetBrains.Annotations
open JetBrains.Application
open JetBrains.Application.Components
open JetBrains.Platform.MsBuildModel
open JetBrains.ProjectModel
open JetBrains.ProjectModel.Model2.Assemblies.Interfaces
open JetBrains.ProjectModel.ProjectsHost
open JetBrains.ProjectModel.ProjectsHost.MsBuild
open JetBrains.ProjectModel.ProjectsHost.SolutionHost
open JetBrains.ProjectModel.Properties
open JetBrains.ProjectModel.Properties.Managed
open JetBrains.ReSharper.Resources.Shell
open JetBrains.Util
open Microsoft.FSharp.Compiler.SourceCodeServices

[<ShellComponent>]
type FSharpProjectPropertiesRequest() =
    let properties = [ "OtherFlags"; "WarnOn" ]
    interface IProjectPropertiesRequest with
        member x.RequestedProperties = properties :> seq<_>

[<SolutionComponent>]
type FSharpProjectOptionsBuilder(solution : ISolution) =
    let msBuildHost = solution.ProjectsHostContainer().GetComponent<MsBuildProjectHost>()
    let defaultDelimiters = [| ';'; ','; ' ' |]

    member x.BuildSingleProjectOptions (project : IProject) =
        let properties = project.ProjectProperties
        let buildSettings = properties.BuildSettings :?> IManagedProjectBuildSettings

        let options = List()
        options.AddRange(seq {
            yield "--out:" + project.GetOutputFilePath(project.GetCurrentTargetFrameworkId()).FullPath
            yield "--simpleresolution"
            yield "--noframework"
            yield "--debug:full"
            yield "--debug+"
            yield "--optimize-"
            yield "--tailcalls-"
            yield "--fullpaths"
            yield "--flaterrors"
            yield "--highentropyva+"
            yield "--target:" + x.GetOutputType(buildSettings)
          })

        options.AddRange(x.GetReferencedPathsOptions(project))

        let definedConstants = x.GetDefinedConstants(properties)
        options.AddRange(List.map (fun c -> "--define:" + c) definedConstants)

        match properties.ActiveConfigurations.Configurations.SingleItem() with
        | :? IManagedProjectConfiguration as cfg ->
            options.Add(sprintf "--warn:%d" cfg.WarningLevel)

            let doc = cfg.DocumentationFile
            if not (doc.IsNullOrWhitespace()) then options.Add("--doc:" + doc)

            let nowarn = x.SplitAndTrim(cfg.NoWarn, defaultDelimiters).Join(",")
            if not (nowarn.IsNullOrWhitespace()) then options.Add("--nowarn:" + nowarn)

            let properties = cfg.PropertiesCollection
            options.AddRange(x.SplitAndTrim(properties.TryGetValue("OtherFlags"), ' '))
        | _ -> ()

        let options' =
            { ProjectFileName = project.ProjectFileLocation.FullPath;
              ProjectFileNames = x.GetProjectFileNames(project)
              OtherOptions = options.ToArray()
              ReferencedProjects = [||]
              IsIncompleteTypeCheckEnvironment = false
              UseScriptResolutionRules = false
              LoadTime = DateTime.Now
              OriginalLoadReferences = []
              UnresolvedReferences = None
              ExtraProjectInfo = None }
        
        options', definedConstants
        
    member private x.GetProjectFileNames(project : IProject) =
        let projectMark = project.GetProjectMark().NotNull()
        let projectDir = projectMark.Location.Directory
        let files = List()
        ignore (msBuildHost.mySessionHolder.Execute(fun session ->
            session.EditProject(projectMark, fun editor ->
                for item in editor.Items do
                    if BuildAction.GetOrCreate(item.ItemType()).IsCompile() then
                        let path = FileSystemPath.TryParse(item.EvaluatedInclude)
                        if not path.IsEmpty then
                            files.Add(x.EnsureAbsolute(path, projectDir).FullPath))))
        files.ToArray()

    member private x.GetReferencedPathsOptions(project : IProject) =
        let framework = project.GetCurrentTargetFrameworkId()
        seq { for p in project.GetReferencedProjects(framework) ->
                  "-r:" + p.GetOutputFilePath(p.GetCurrentTargetFrameworkId()).FullPath
              for a in project.GetAssemblyReferences(framework) ->
                  "-r:" + a.ResolveResultAssemblyFile().Location.FullPath }

    member private x.EnsureAbsolute(path : FileSystemPath, projectDirectory : FileSystemPath) : FileSystemPath =
        let relativePath = path.AsRelative()
        if isNull relativePath then path
        else projectDirectory.Combine(relativePath)

    member private x.GetOutputType([<CanBeNull>] buildSettings : IManagedProjectBuildSettings) =
        if isNull buildSettings then "library"
        else
            match buildSettings.OutputType with
            | ProjectOutputType.CONSOLE_EXE -> "exe"
            | ProjectOutputType.WIN_EXE -> "winexe"
            | ProjectOutputType.MODULE -> "module"
            | _ -> "library"

    member private x.GetDefinedConstants(properties : IProjectProperties) =
        match properties.ActiveConfigurations.Configurations.SingleItem() with
        | :? IManagedProjectConfiguration as cfg -> x.SplitAndTrim(cfg.DefineConstants, defaultDelimiters)
        | _ -> List.empty

    member private x.SplitAndTrim([<CanBeNull>] strings, [<ParamArray>] delimiters : char[]) : string list =
        if isNull strings then List.empty
        else [ for s in strings.Split(delimiters) do if not (s.IsNullOrWhitespace()) then yield s.Trim() ]